### PR TITLE
Burninate `let`

### LINKED ISSUE #180: Macro expansion works differently from Lisp conventions
Given

```
(defmacro rev [&rest body]
  (quasiquote (do (unquote-splice (list (reversed body))))))
```

(shamelessly stolen from native-tests), the expression

```
(let [[rev math.sqrt]] (rev 2))
```

evaluates to `2` whereas in other Lisps I know it evaluates to `(sqrt 2)`. What's worse is that other Lisps' behaviour is more useful. Is this a bug or a feature?

### LINKED ISSUE #576: `let` leaves variables set even after leaving scope
I was playing around with the idea of porting Clojure's [`with-redefs`](http://clojuredocs.org/clojure_core/clojure.core/with-redefs) macro, and I thought for a moment I might be able to get away with jut using `let`.  To my dismay:

```
=> (import [numpy :as np])
=> (np.cos 2)
-0.41614683654714241
=> (let [[foocos (fn [x] 1)] [np.cos foocos]] (np.cos 2))
1L
=> (np.cos 2)
1L
```

This doesn't seem right... shouldn't `let` return `np.cos` to its former value when `np.cos` leaves scope?

### LINKED ISSUE #619: yield within let within loop does not work
I'm not completely sure if this is a bug or expected behaviour. At the very least, it's a confusing gotcha.

```
(defn test []
  (while true
    (let [[foo "bar"]]
      (yield foo))))
```

produces the following python code:

```
def test():
    while True:

        def _hy_anon_fn_1():
            foo = u'bar'
            yield foo
        _hy_anon_fn_1()
```

which results in an infinite loop of creating the generator object, instead of returning it from test.

Exchanging positions of the let and while provides the expected result:

```
(defn test []
  (let [[foo "bar"]]
    (while true
      (yield foo))))
```

```
def test():

    def _hy_anon_fn_1():
        foo = u'bar'
        while True:
            yield foo
    return _hy_anon_fn_1()
```

### LINKED ISSUE #663: Yield does not work inside nested `let` and `if`
Such code:

``` lisp
(let [[var 1]]
     (if var
         (yield "Hello")))
```

Compiles into:

``` python
def _hy_anon_fn_1():
    var = 1L
    return (yield u'Hello' if var else None)
_hy_anon_fn_1()
```

And return inside a generator leads to `SyntaxError: 'return' with argument inside generator` error.

### LINKED ISSUE #664: `Let` creates a closure where I unable to change outer variables because of UnboundLocalError
``` lisp
(defun test []
  (setv foo true)
  (let []
       (if foo
           (setv foo false))))

(test)
```

Compiles into:

``` python
def test():
    foo = True

    def _hy_anon_fn_1():
        if foo:
            foo = False
            _hy_anon_var_1 = foo
        else:
            _hy_anon_var_1 = None
        return _hy_anon_var_1
    return _hy_anon_fn_1()
test()
```

which leads to:

```
Traceback (most recent call last):
  File "/home/art/yandex-certificator/hy/env/bin/hy", line 8, in <module>
    load_entry_point('hy==0.10.0', 'console_scripts', 'hy')()
  File "/home/art/yandex-certificator/hy/hy/cmdline.py", line 305, in hy_main
    sys.exit(cmdline_handler("hy", sys.argv))
  File "/home/art/yandex-certificator/hy/hy/cmdline.py", line 293, in cmdline_handler
    return run_file(options.args[0])
  File "/home/art/yandex-certificator/hy/hy/cmdline.py", line 194, in run_file
    import_file_to_module("__main__", filename)
  File "/home/art/yandex-certificator/hy/hy/importer.py", line 74, in import_file_to_module
    eval(ast_compile(_ast, fpath, "exec"), mod.__dict__)
  File "minimal", line 7, in <module>
    (test)
  File "minimal", line 3, in test
    (let []
  File "minimal", line 4, in _hy_anon_fn_1
    (if foo
UnboundLocalError: local variable 'foo' referenced before assignment
```

I know why this does not work in python, but this behaviour is not expected in lisp like language.

### LINKED ISSUE #703: Invalid "SyntaxError: 'break' outside loop"
The following code should compile correctly but the compiler instead says that there is an invalid break outside of a loop, which is not true:

=> (while true
  (let ((y false)) 
    (if (not y) 
    (break) 
      (setv y 1))))
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/hy/importer.py", line 42, in ast_compile
    return compile(ast, filename, mode, flags)
  File "<input>", line 1
SyntaxError: 'break' outside loop

### LINKED ISSUE #1000: Re-implement let
Feedback on the following:

``` hy
(defmacro let [forms &rest body]
  (setv closure-name (gensym "CLOSURE"))
  (setv closure `(do (setv ~closure-name {})))

  (defn generic-visit [visit node]
    (if (coll? node)
        (setv (cut node) (list-comp (visit x) [x node]))
        node))

  (defn name-mangle [forms]
    (setv names (list-comp (get x 0) [x forms]))
    (defn visit [node]
      (if (in node names)
          `(get ~closure-name (quote ~node))
          (generic-visit visit node))))

  `(do ~closure  ; assign {} to something like ':CLOSURE_35151
       ; Now, let's assign all the bindings to the closure we know about
       ~@(list-comp `(assoc ~closure-name (quote ~name) ~value) [(, name value) forms])
       ; And, now, let's change any names to ref the dict.
       ~@(map (name-mangle forms) body)))
```

This fucks up locals when using gdb, but it might work for trying to actually write a LoL

### LINKED ISSUE #1055: Get rid of break/continue and restrict yield to new defgn/gn special form
If we could make `yield`, `break`, or `continue` in a `let`-form a syntax error, then `let` would almost work properly. Unfortunately, detecting this correctly in all cases seems _really_ difficult.

With that done, the only remaining potential issue would be a local assignment shadowing a nonlocal. I don't think this is such a big deal, since Python can _read_ nonlocals just fine and Clojure seems to get along fine without writable locals period. We could work around this the same way Clojure(script) does--with an atom construct. We can even use explict `nonlocal` statements instead in Python 3.

It would be easier to get `let` working if we got rid of `yield`, `break` and `continue`. But maybe we don't have to go _quite_ that far.
#### Get rid of `break`/`continue`

It seems contradictory to me that we don't have `return`, and yet allow `break` and `continue`, since the main arguments against `return` apply equally to them, and Python uses explicit returns to compensate for its lack of labels anyway. To be consistent, why not get rid of them too?

We could replace them with forms that don't compile to Python's `break`/`continue`, but throw exceptions instead. A `LabelBreak` exception would still work in infinite loops, and would work properly inside `let` as implemented now. The required `try`/`except` could be hidden with a required explicit `label` macro.

``` Hy
(label :x
  (while 1
    (let []
      (break :x))))
```

Which becomes something like:

``` Python
try:
  while 1:
    def _hy_anon_fn_1():
      raise LabelBreak('\ufdd0:x')
except LabelBreak as :G_1234:
  if :G_1234.is_not_named('\ufdd0:x'):
    raise
```

Getting `continue` behavior is as simple as moving the label inside the loop:

``` Hy
(while (foo)
  (label :y
    (let []
      ;; do stuff?
      (break :y))))
```

Which becomes:

``` Python
while foo():
  try:
    def _hy_anon_fn_1():
      # do stuff?
      raise LabelBreak('\ufdd0:y')
  except LabelBreak as :G_1234:
    if :G_1234.is_not_named('ufdd0:y'):
      raise
```

Since they re-raise exceptions that don't match the label, there's no reason you can't use both inner and outer labels when required. You can also break out of nested loops. `try` forms can also have an `else` block like Python loops do, so any `else` logic the loop needs could be moved to the `label` form itself.

There's actually no reason that break logic has to be attached to a loop, and this `label/break` construct cleanly separates the ideas. It works just like Java's [labeled block statements](https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.7); a more structured alternative to a full-on `goto` like in C.

This will also compensate for the lack of `return`.

``` Hy
(defn foo []
  (label :return
    ;; do stuff?
    (setv res something)
    (break :return))
  res)
```

We could even build this behavior into the label form, so that `(label :x (break :x res))` has a value of `res`. See [Drython's statement module](https://github.com/gilch/drython/blob/master/drython/statement.py) for more examples of how this could work.
#### Restrict `yield` to `gn`

`yield` is such a powerful feature of Python, and so important to coroutintes, that I can't imagine getting rid of it is a good idea. But maybe we can make it behave better in Hy.

A generator function is a very different beast in Python from a normal function `def`, with its own quirks and rules. Fewer special cases make macro writing easier. They would be easier to deal with in Hy, both in `let` and other macros, if we cleanly separated them.

Rather than trying to detect the presence of `yield` in the `let` macro, I propose adding a new special form `gn` (for generator) analogous to `fn`. It would also have an equivalent `defgn` macro.

 The ast generator will always add the lines

``` Python
    return
    yield
```

to the end of a `gn`'s `def` expansion. This guarantees that Python will always treat it as a generator function (since it contains a `yield`), even if it's empty (since it always returns before it reaches the final `yield`), but won't change the behavior of any code above it in the `def`.

If you need a generator, you can't use `fn`/`defn` with `yield` anymore. Use `gn`/`defgn` instead.

Hy's ast generator must enable `yield` when inside a `gn` special form, _and disable it_ again when inside a `fn` special form, especially if a `fn` is nested inside a `gn`. This way, we can leave `let` as-is, that is, defined in terms of a `fn`, which now can't possibly contain a `yield` to muck things up. Other macros would also be free to use `let` in their expansions in the traditional Lisp manner without the same worry.

### LINKED ISSUE #1056: remove let proposal (again!) and maybe replace with deprecated contrib shim
#### Proposal

Remove `let` from core Hy. Redefine any library functions/macros in terms of `setv` if possible, or `fn` if it requires a closure.
##### Sub-Proposal

As a compromise, to make porting old code easier, implement a deprecated `let` macro in contrib, with dire warnings that it may be removed in future versions of Hy, and an explanation of why it's broken. We could even have different versions with double brackets or not and with a `((fn ...))` or `do`/`setv` expansion.
#### Rationale

Keeping `let` as-is requires us to remove or replace the the things it breaks, namely `break`/`continue`, `yield`, and now `async`/`await` #1054 (and possibly any other future Python syntax). I think I can replace `break` and `continue` #1055, but nobody knows how to replace `yield`/`async`/`await`. Removing them is out of the question. I'd sooner give up `let` than coroutines, wouldn't you?

Keeping the things `let` breaks requires us to remove or replace `let`. Nobody's figured it how to fix `let` after a year, not that we haven't tried #1055 #1000 #844 #688 #673 #664 #663, and this problem has all but deadlocked Hy development in that time.

Furthermore, if performance is a concern (it's not so much for me), then it's much more efficient in Python to set locals than to declare a function, call the function, and set locals. Most Python implementations don't inline functions (last I checked) yet `let` is almost exclusively used for lexically-scoped locals now.

If we _don't_ have `let` in core, and IF we get it working eventually, we can put it back in without breaking API again. Accepting this proposal leaves us that hope, but if we settle for any version of `let` that doesn't quite work right, we can't do that.

---

I know that my original proposal #844 got closed with the loose consensus that we'd replace `let` with a working version. After banging our heads against this for about a year with no success, perhaps some of us have changed our minds?

The rules say it only takes three to push through a change, one to commit, and two (from core) to approve. @kirbyfan64, @tuturto, what say you? Of course, three more could revert it, so I'll call that a veto. @hylang/core, objections? Vetoes? Approval? Counter proposals? Crazy ideas?

### LINKED ISSUE #1212: `let` can lead to avoidable unbound local error
On the surface this might look like #664 but this would be easily avoided if `let` is implemented closer to Lisp (or was it Scheme?)

```clojure
(defn test [x]
  (let [x (inc x)]
    (print x)))
```

Compiled to:

```python
def test(x):
    def _hy_anon_fn_1():
        x = inc(x)
        return print(x)
    return _hy_anon_fn_1()
```

But as I recall `let` in List/Scheme being just a transformation to a anonymous function immediately called I expected *this* as equivalent Python code:

```python
    def test(x):
        def _hy_anon_fn_1(x):
            return print(x)
        return _hy_anon_fn_1(inc(x))
```

That doesn't have the unbound local problem. Why isn't `let` transformed this way?

### PR DESCRIPTION
**Pull request by Kodiologist**
See the discussion in #1056.

I haven't touched `scripts/reformat-changelog`, which still has some uses of `let`, because I'm not totally sure what it's supposed to do, so I wouldn't know how to check that I didn't break it.

- Closes #180 (I think)
- Closes #576
- Closes #619
- Closes #663
- Closes #664
- Closes #703
- Closes #1000
- Closes #1055
- Closes #1056
- Closes #1212

And maybe some other `let`-related bugs I didn't catch.

### CODE REVIEW

**Comment by Kodiologist** [comment]
Travis says something's broken on Python 2. I'll look into it.

File: docs/language/api.rst [hunk 1]
```diff
@@ -286,17 +286,17 @@ Examples of usage:
 
 .. code-block:: clj
 
-  =>(let [collection {}]
+  =>(do (setv collection {})
```
Comment (kirbyfan64) [inline docs/language/api.rst]: Should the indentation on these be adjusted now to line up better?

File: hy/core/bootstrap.hy [hunk 2]
```diff
@@ -41,23 +41,13 @@
 
 (defmacro defn [name lambda-list &rest body]
   "define a function `name` with signature `lambda-list` and body `body`"
-  (if (not (= (type name) HySymbol))
+  (import hy)
```
Comment (kirbyfan64) [inline hy/core/bootstrap.hy]: As a side note: why exactly was this changed?

File: tests/native_tests/language.hy [hunk 3]
```diff
@@ -1037,20 +993,6 @@
   (assert (is (xor 0 False) False)))
 
 
-(defn test-if-return-branching []
```
Comment (kirbyfan64) [inline tests/native_tests/language.hy]: How much of this test was actually do to `let`? It was added in [this bugfix commit](https://github.com/hylang/hy/commit/5336b2f71ace09ab3df9af56238eb16289f0aafa), yet the bugfix didn't seem to have anything to do with let. Was this fixing another bug that just *happened* to be easy to reproduce using `let`?

File: hy/core/bootstrap.hy (hunk 2 above)
Comment (Kodiologist) [inline hy/core/bootstrap.hy]: See the description of  6efe918.

File: tests/native_tests/language.hy (hunk 3 above)
Comment (Kodiologist) [inline tests/native_tests/language.hy]: I don't know. I can't get a version of Hy that old to run in order to try to reproduce the bug without `let`.

File: docs/language/api.rst (hunk 1 above)
Comment (Kodiologist) [inline docs/language/api.rst]: Done.

File: docs/language/internals.rst [hunk 4]
```diff
@@ -426,10 +429,10 @@ so our re-written ``nif`` would look like:
 
    (defmacro nif [expr pos-form zero-form neg-form]
```
Comment (tuturto) [inline docs/language/internals.rst]: example is broken. A working version could be:
```hy
(defmacro nif [expr pos-form zero-form neg-form]
  (with-gensyms [g]
    `(do
       (setv ~g ~expr)
       (cond [(pos? ~g) ~pos-form]
             [(zero? ~g) ~zero-form]
             [(neg? ~g) ~neg-form]))))
```

Comment (Kodiologist) [inline docs/language/internals.rst]: The version of `nif` you commented on here seems to work for me, both with this branch (`let-it-be`) and with master.

Comment (tuturto) [inline docs/language/internals.rst]: Yeah, looks like it. My bad. I must have done something extremely silly while testing that.

**Comment by Kodiologist** [comment]
@kirbyfan64 We good?

Incidentally, while you're changing your username and stuff, I miss your old avatar that actually had Kirby in it.

**Comment by kirbyfan64** [comment]
@Kodiologist Haha, I actually changed my username, but then I changed it back. Too many memories. :)

**Comment by kirbyfan64** [comment]
Also, this gets my :+1:, but would you mind letting me check out that test a sec first? Just to be on the safe side!

**Comment by Kodiologist** [comment]
Sure, let me know when you're done (or do the merge yourself).

File: tests/native_tests/language.hy (hunk 3 above)
Comment (kirbyfan64) [inline tests/native_tests/language.hy]: @Kodiologist Got it. Here's a test case that still applies to the latest Hy:

```hy
(defn test-if-return-branching []
  "NATIVE: test the if return branching"
                                ; thanks, algernon
  (defn f []
    (if true (setv x 1) 2)
    1)

  (assert (= 1 (f))))
```

Comment (Kodiologist) [inline tests/native_tests/language.hy]: Great, thanks. I'll rebase and merge.

**Comment by Kodiologist** [comment]
TFW you close 10 issues with one merge.

**Comment by svetlyak40wt** [comment]
@Kodiologist scripts/reformat-changelog was intended to reformat NEWS file into more conventional ChangeLog in debian format, where a version number is followed by it's description because a line like `Changes from 0.12.1` says nothing about what current version is. Whether it is 0.13.0 or 0.12.2.

That was the purpose – to use this script – reformat changelog and drop NEWS file.

But I see, nobody cares except me :(

**Comment by Kodiologist** [comment]
@svetlyak40wt In that case, it seems like we shouldn't be tracking reformat-changelog in the repository; you can just open a PR to replace the NEWS file with a ChangeLog, and we'll update the ChangeLog directly henceforth.

**Pull request merged**